### PR TITLE
fix: use forgiving jwt authentication

### PR DESCRIPTION
**Pull request by robrap**
**Description:**

The idea behind this work is to use "forgiving" jwt authentication
when working with JWT cookies, which means that if jwt cookie
authentication fails, we would try other authentication mechanisms,
rather than failing and halting the authentication attempt.

This solution would replace an earlier solution to issues around
jwt cookies that introduced an HTTP_USE_JWT_COOKIE header.

For more background and details, see the new ADR:
0002-remove-use-jwt-cookie-header.rst.

Also contains:
* add custom attributes:
  * is_forgiving_jwt_cookies_enabled
  * use_jwt_cookie_requested
  * has_jwt_cookie
  * jwt_auth_result
* ENABLE_FORGIVING_JWT_COOKIES toggle for rollout of what would
  be a future breaking change, when we have fully switched to
  forgiving jwt cookies.

**JIRA:**

- DEPR: https://github.com/openedx/edx-drf-extensions/issues/371
- Legacy ticket: [ARCHBOM-1218](https://openedx.atlassian.net/browse/ARCHBOM-1218)

**Author concerns:**
 
- Ecommerce would be the riskiest rollout since it is using all affected Middleware.

**Reviewers:**
- [ ] tag reviewer

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bump if needed
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/authentication.py:81]: **Note to Reviewer:** As noted in the comment, this case was not appearing in New Relic. Instead, the below case is sometimes reporting: `Exception:AuthenticationFailed('Invalid token.')`. This case was removed to simplify the code, including the "original" code.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:295]: **Note to Reviewer:** It seems like this change would work even in the "original" case, and would potentially be more accurate. It could be moved to a separate PR (also with a rollout flag?), or just left here as part of the larger rollout.

File: edx_rest_framework_extensions/middleware.py
Comment (robrap) [inline edx_rest_framework_extensions/middleware.py:158]: **Note to Reviewer:** Since the consolidated jwt cookie with `jwt_cookie_name()` wouldn't exist if `USE_JWT_COOKIE_HEADER` didn't exist, it seems like this case was never needed.

File: requirements/base.in
Comment (robrap) [inline requirements/base.in:9]: **Note to Self:** Remove this. Instead of using a `SettingToggle`, I just used a Django Setting directly.

**Review by robrap — COMMENTED** [review]
Notes for my future self or future reviewers. Also see "Author Concerns" of the PR description.

**Comment by robrap** [comment]
Closing this draft PR with comments for ARCHBOM-1218 until the ticket is picked up again.

**Comment by feanil** [comment]
Re-opening this because I think it's still relevant once we update it a bit.  I'll take a stab at re-basing it off of the latest main branch.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:269]: @feanil: I noticed this in the logs recently. There are a bunch of `log.warning`s here that should be login.info at most, and probably even log.debug. What do you think?

File: requirements/base.in
Comment (robrap) [reply requirements/base.in:9]: Delete this line:
```suggestion
```

File: edx_rest_framework_extensions/config.py
Comment (robrap) [inline edx_rest_framework_extensions/config.py:23]: It would be great to add a ticket for cleanup of the toggle.

File: docs/decisions/0002-remove-use-jwt-cookie-header.rst
Comment (robrap) [inline docs/decisions/0002-remove-use-jwt-cookie-header.rst:30]: 1. I'm removing this comment, because I don't find it useful. This change is encapsulated to this library, so this is not a backward incompatible change.
```suggestion
```
2. I did find one bit of cleanup in an edx-platform test that can come when we remove the temporary toggle and this header is no longer used. See https://github.com/openedx/edx-platform/blob/f0a9ef21613a780e4c7975de47375fe7f21e3710/openedx/core/djangoapps/user_authn/tests/test_cookies.py#L78.
3. We need a minor version bump and changelog entry for this change.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (feanil) [inline edx_rest_framework_extensions/auth/jwt/authentication.py:110]:
```suggestion
        If the JWT is in the authorization header, and authentication fails, we will raise
        an exception which will halt additional authentication methods, and is the default
        behavior of DRF authentication classes.
```

File: docs/decisions/0002-remove-use-jwt-cookie-header.rst
Comment (feanil) [reply docs/decisions/0002-remove-use-jwt-cookie-header.rst:30]: Are you saying it's backwards compatible because we can roll back the library version of because of some other reason?

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (feanil) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:225]:
```suggestion
        Additionally, may add the user to the request to make it available in process_view. (See below.)
```

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (feanil) [reply edx_rest_framework_extensions/auth/jwt/middleware.py:269]: That makes sense, it sounds like that would be a follow-up ticket in this repo? Is making an issue for this repo right now, the correct next step?

File: edx_rest_framework_extensions/config.py
Comment (feanil) [inline edx_rest_framework_extensions/config.py:22]:
```suggestion
# .. toggle_creation_date: 2023-05-17
# .. toggle_target_removal_date: 2023-06-31
```

**Review by feanil — COMMENTED** [review]
Just a few comments from a more thorough review, next I'm gonna try to get this going locally and see what the test coverage looks like.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/authentication.py:110]: Curious if we have a style documentation around this? I think I was introduced to the indenting by Régis, but could have been someone else, and I prefer the indents. It means multi-paragraph comments don't need blank lines, which clarifies that they are grouped. Non-blocking either way, but just curious if this is simply competing preferences.

File: docs/decisions/0002-remove-use-jwt-cookie-header.rst
Comment (robrap) [reply docs/decisions/0002-remove-use-jwt-cookie-header.rst:30]: I mean that upgrading to this version of the library _shouldn't_ break anything, including when the new setting is toggled on.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/middleware.py:269]: Ticket sounds right.

File: docs/decisions/0002-remove-use-jwt-cookie-header.rst
Comment (robrap) [inline docs/decisions/0002-remove-use-jwt-cookie-header.rst:48]: @feanil: Do you plan on ticketing other follow-up work like this as well?

Comment (robrap) [reply docs/decisions/0002-remove-use-jwt-cookie-header.rst:48]: It looks like it is `USE-JWT-COOKIE` (and `USE_JWT_COOKIE`), so a more complete search is https://github.com/search?q=%28org%3Aopenedx+OR+org%3Aedx%29+%28USE_JWT_COOKIE+OR++USE-JWT-COOKIE%29&type=code. It looks like a bunch of stuff in the edx repo that will needs clean-up at some point.

**Comment by robrap** [comment]
@feanil: What's your plan for this?

**Comment by feanil** [comment]
@robrap I had to backburner this a bit to get some other work around documentation moving.  I think I'll be able to finish that up this week and then I can come back to this next week.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/authentication.py:101]: Thoughts on monitoring:
1. Can we add annotations to the various set_custom_attribute calls? (search code base for custom_attribute_name)
2. For the forgiving JWT case, would it be better to rename this to something like jwt_auth_failed_warning? Having one name (no change) may make this simple to find, but it may also be misleading if someone doesn’t know that forgiving jwt custom attributes need to be reviewed as well to know how to interpret.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/authentication.py:73]: 1. I feel like we should add `enabled` to the name, if we keep this attribute, to clarify this is just about the toggle. This is a long name though.
2. I’m wondering if it would be better to have an additional or replacement attribute like jwt_auth_failure_forgiven, which would be True when this new functionality actually has an effect.

**Review by robrap — COMMENTED** [review]
Adding some monitoring thoughts while I’m thinking of it.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (feanil) [reply edx_rest_framework_extensions/auth/jwt/middleware.py:273]: Created https://github.com/openedx/edx-drf-extensions/issues/363

**Comment by feanil** [comment]
@robrap I think you specifically are going to be the one looking at this data so I'd love something more concrete about any changes you might like to the monitoring keys or values to make that easy for you.  Your thoughts seem reasonable but it's unclear if you want me to take any actions to make your life easier.

That said, I think this PR is ready to land, We've got test coverage of the new code and i've tested that if I do flip the forgiving JWT default in the settings in this repo, all tests here pass.  I think this is ready to merge and publish once we work out the details about the monitoring.

The test failure is just CodeCov upload failing 🙄 hopefully it will work in a little bit.

File: docs/decisions/0002-remove-use-jwt-cookie-header.rst
Comment (robrap) [reply docs/decisions/0002-remove-use-jwt-cookie-header.rst:30]: @feanil: Can we commit my suggestion (to delete text) above, or do you still have a question about this?

File: docs/decisions/0002-remove-use-jwt-cookie-header.rst
Comment (robrap) [reply docs/decisions/0002-remove-use-jwt-cookie-header.rst:48]: @feanil: Do we have a ticket for this? Can you add a link? Another point about the rollout is that we can't land the removal of the old header until we ensure that **all** backends are using forgiving cookies. We might be able to remove the toggle before that, but it may not be worth it.

File: edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py:252]: There are some issues with this testing, but not certain how deep:
1. The final test also overrides `EDX_DRF_EXTENSIONS`, and based on codecov message, it looks like `ENABLE_FORGIVING_JWT_COOKIES` may not actually get overridden as expected. Maybe we just dupe that one test?
2. Not sure if we want to override the tests again, but:
  a. We might want a conditional around `request.META[USE_JWT_COOKIE_HEADER] = 'true'` based on `ENABLE_FORGIVING_JWT_COOKIES`, and
  b. Ideally we'd test forgiving JWTs with and without this header, to ensure it will work both before and after the header is removed. However, since you've been testing with it, if we add the conditional we'll have switched to testing without, and hopefully that is enough.

File: edx_rest_framework_extensions/config.py
Comment (robrap) [inline edx_rest_framework_extensions/config.py:22]:
```suggestion
# .. toggle_creation_date: 2023-08-01
# .. toggle_target_removal_date: 2023-10-01
```

File: edx_rest_framework_extensions/config.py
Comment (robrap) [reply edx_rest_framework_extensions/config.py:23]: @feanil: Can you ticket and add a link?

File: edx_rest_framework_extensions/config.py
Comment (robrap) [reply edx_rest_framework_extensions/config.py:22]: I am marking this as resolved because I added a new comment with newer dates.

File: docs/decisions/0002-remove-use-jwt-cookie-header.rst
Comment (robrap) [reply docs/decisions/0002-remove-use-jwt-cookie-header.rst:48]: Note that we might just need one rollout ticket with various steps. See this old comment as well: https://github.com/openedx/edx-drf-extensions/pull/197#discussion_r1190058907

File: .editorconfig
Comment (robrap) [reply .editorconfig:1]: @feanil: I still think this file should be ignored and removed from the PR. Thanks.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/authentication.py:144]:
```suggestion
                # Returning None is how a JWT cookie failure becomes forgiving.
                # See docs/decisions/0002-remove-use-jwt-cookie-header.rst for details.
```

File: edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py:370]: 1. Typo:
```suggestion
class TestForgivingJwtRedirectToLoginIfUnauthenticatedMiddleware(TestJwtRedirectToLoginIfUnauthenticatedMiddleware):  # pylint: disable=test-inherits-tests # noqa E501 line too long
```
2. I think I don't see any of the same issues that I saw with the other duplicated test class, but not certain.

File: edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py:503]: For this duplicate class, I think I see the same issues detailed in my other comment.

File: edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py:193]: Although this is possibly good enough, once we commit to the custom attributes, it might be good to verify that we have captured the CSRF failure.

**Review by robrap — COMMENTED** [review]
Thanks @feanil. I'm going to discuss monitoring with someone on my team late tomorrow, and will come back with a proposal around that. In the meantime, I've noted other things.

Also, when done, please check the Conversations tab to ensure you caught all the comments. Some of my comments are not new, but just pings on old comments. Thanks!

File: edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py
Comment (feanil) [reply edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py:252]: So the existing modifications had to be made because of the case where we don't pass the header.  Previously those test were failing but with forgiving JWT auth, they begin passing so that case is already covered.

For the code coverage, I think there is some issue  with the latest coverage report being generated and sent to codecov. When I looked at the coverage locally, I was able to confirm that both code paths were getting covered after I duplicated the tests and added the override on the duplicates.

File: edx_rest_framework_extensions/config.py
Comment (feanil) [inline edx_rest_framework_extensions/config.py:23]:
```suggestion
# .. toggle_tickets: https://github.com/openedx/edx-drf-extensions/issues/365
```

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (feanil) [reply edx_rest_framework_extensions/auth/jwt/authentication.py:73]: I'll make these updates, today or tomorrow.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (timmc-edx) [inline edx_rest_framework_extensions/auth/jwt/middleware.py:236]: Suggested recombination of the two methods: https://github.com/openedx/edx-drf-extensions/pull/366

**Review by robrap — COMMENTED** [review]
@feanil: Confirming that I will be updating the refactor PR and fixing some of the monitoring there, and then we can merge back in to this PR and finish any remaining issues here.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/authentication.py:69]: @timmc-edx: [inform] I'm not sure if this will be temporary, so I did **not** mark it as such.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/authentication.py:73]: I'm taking care of this. I'll commit changes next week when I return.

File: docs/decisions/0002-remove-use-jwt-cookie-header.rst
Comment (robrap) [reply docs/decisions/0002-remove-use-jwt-cookie-header.rst:46]: Marking this resolved. Here is the ticket: https://github.com/openedx/edx-drf-extensions/issues/365

File: edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py:216]: Marking as resolved.

File: edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py:286]: I'm marking as resolved. I don't plan on adding a whole other set of tests, and if we rely on that header when we don't think we should, we'll find out when we go to remove it.

File: edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py:503]: Feanil explained elsewhere. Marking as resolved.

File: edx_rest_framework_extensions/config.py
Comment (robrap) [reply edx_rest_framework_extensions/config.py]: Feanil had updated this. Marking as resolved.

File: .editorconfig
Comment (robrap) [reply .editorconfig:1]: I now understand that this is first time add from linting. Marking this as resolved.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [inline edx_rest_framework_extensions/auth/jwt/authentication.py]: **Note to reviewer:** The refactor makes a change to this original code (not seen in the changed code) where
we now check to see if there is JWT cookie, rather than checking if the `USE_JWT_COOKIE` header exists, to determine whether to check for CSRF protection.

**Review by robrap — COMMENTED** [review]
@timmc-edx @feanil: You may want to review the Aug 8th and Aug 10th commits by commit.

Here is a summary of changes:

- docs: update annotations for request_auth_type_guess 
- docs: add annotations for jwt_auth_failed
- fixup! forgiving jwt custom attribute updates
    - replace jwt_auth_has_jwt_cookie (was in JwtAuthentication) with new has_jwt_cookie in JwtAuthCookieMiddleware
    - refactor and rename jwt_auth_forgiving_jwt_cookies to is_forgiving_jwt_cookies_enabled
    - annotate is_forgiving_jwt_cookies_enabled
    - add and annotate jwt_auth_failure_forgiven
        - Note: This is replaced by jwt_auth_result in a future commit.
- fixup! refactor JwtAuthentication authenticate (combined orginal and forgiving_jwt_cookies)
    - Note: this was another refactor like the one from the separate PR
- feat: enable JWT cookie testing
    - adds a test setting used in a follow-up commit
- fixup! switch to jwt_auth_result custom attribute
    - replaces jwt_auth_failure_forgiven

**UPDATE**: Also added a changelog entry and bumped the version.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/authentication.py]: See https://github.com/openedx/edx-drf-extensions/pull/197#pullrequestreview-1572779330 for changes. Marking as resolved.

File: edx_rest_framework_extensions/auth/jwt/middleware.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/middleware.py]: Done.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/authentication.py]: See https://github.com/openedx/edx-drf-extensions/pull/197#pullrequestreview-1572779330 for changes. Marking as resolved.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/authentication.py]: See https://github.com/openedx/edx-drf-extensions/pull/197#pullrequestreview-1572779330 for changes. Marking as resolved.

**Review by feanil — APPROVED** [review]
One question, otherwise looks good to me.

File: edx_rest_framework_extensions/auth/jwt/authentication.py
Comment (robrap) [reply edx_rest_framework_extensions/auth/jwt/authentication.py]: @feanil: You are correct. Also, this comment went away when this attribute was removed in favor of jwt_auth_result, so there is no doc updates needed.

**Pull request merged**
